chain/epoch-manager/src/shard_tracker.rs

```diff
@@ -49,9 +56,29 @@ impl ShardTracker {
         match &self.tracked_shards_config {
             TrackedShardsConfig::NoShards => Ok(false),
             TrackedShardsConfig::AllShards => Ok(true),
+            TrackedShardsConfig::Shards(tracked_shards) => {
+                let shard_layout = self.epoch_manager.get_shard_layout(epoch_id)?;
+                if !shard_layout.shard_ids().contains(&shard_id) {
+                    return Err(EpochError::ShardingError(format!("Invalid shard id {shard_id}")));
+                }
```

Comment:
I wonder if this could be a debug assert and if it could be extracted before the match - so that it applies to all variants?

Reply:
Let me create an issue to handle this in a separate PR.
There are methods like `will_care_about_shard` that return true if all shards are tracked, without calling `tracks_shard_at_epoch`.
https://github.com/near/nearcore/pull/13437 consolidates all such `care_about_shard` methods into a single function, so we could add the check there.

Currently, the check would fail in the case of resharding: `get_catchup_and_state_sync_infos` would indirectly call `tracks_shard_at_epoch` with the parent shard ID and the post-resharding epoch ID.

Reply:
Created issue: https://github.com/near/nearcore/issues/13445